fix(panelv2): update tests to use a11y matchers and user events

jendowns opened the pull request:
## Affected issues

- ~BLOCKED BY #308~ No longer blocked!
- follow up to #308 

## Proposed changes

- improve `PanelV2` tests by adding a11y tests + checking for user events like clicking to close the panel & tabbing through panel content

jendowns commented:
@diego-codes I just added you as a reviewer, if you have time --

Keeping in mind that current there are two DAP violations right now and that #308 will ultimately help with that, I was wondering if you could take a look at these tests and let me know if you agree with the structure etc? And if you think there are gaps? Thank you so much 🙏

diego-codes commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,121 @@
 describe('PanelV2', () => {
-  it('renders', () => {
-    const wrapper = mount(
+  afterEach(cleanup);
```
I'm pretty sure you don't need to do clean up anymore. The package does this already.

diego-codes commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,121 @@
+    );
+
+    userEvent.click(getByLabelText(/test close/i));
+    expect(onCloseMock).toHaveBeenCalled();
```
Sometimes I take the assertion a bit further by using `toHaveBeenCalledTimes(1)` just in case so it doesn't call it multiple times.

diego-codes commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,121 @@
       </PanelV2>
     );
-    expect(wrapper).toMatchSnapshot();
+    await expect(document.body).toMatchSnapshot();
```
For this test, what exactly are you wanting to test for that can't be achieved with specific assertions or just looking at a storybook story? The reason I'm asking is that snapshots are so challenging to maintain and identify when real bugs appear. We've started to move away from them.

jendowns replied:
@diego-codes I was keeping this one out of habit, to be honest. 😅 But you make a good point so I'll go ahead and remove it.

SimonFinney replied:
@diego-codes I'd like to understand more about your decisions to move away from snapshot tests if you wouldn't mind elaborating?

I personally still find value from them as long as they're being used in moderation and in conjunction with a more comprehensive testing suite, but I'd love to hear more about the challenges you've faced with them!

diego-codes commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,121 @@
+      }
+    );
+    await expect(document.body).toHaveNoAxeViolations();
+    await expect(document.body).toHaveNoDAPViolations('PanelV2');
```
I'm not familiar with this assertion what does this line mean?

jendowns replied:
I added Axe and DAP matchers in this PR, based on similar matchers Carbon added a month or so ago: https://github.com/carbon-design-system/ibm-security/pull/312
Carbon's matchers: https://github.com/carbon-design-system/carbon/tree/ac14f81df750dde81d125e6a2a036e476723b5b9/tasks/jest/matchers

diego-codes replied:
Oh, I didn't catch that they were two different assertions. Nice! Looks useful.

diego-codes commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,121 @@
+    expect(queryByText(/test title/i)).not.toBeInTheDocument();
+  });
+
+  it('should cycle panel elements in tab order', () => {
```
Does this actually test `PanelV2` logic or is it testing some focus trap component or tool that's not actually part of the panel code? If its the latter, it might be more appropriate to test the focus trap component if it is an internally developed one.

jendowns replied:
This is meant to test tab order, but it wouldn't be able to effectively test focus trap anyway because there's no interactive elements outside of this panel in the mock/render. 🤔

diego-codes replied:
Do you expect the tab order to work any differently than the default browser behavior? If not, this test is testing jsdom instead of any panel logic. 

It looks like the `Portal` component relies on a package called `focus-trap-react` so my answer would not test for focus trapping. However, I see that the trap is activated by the `focusTrap` prop on the `Portal` component so that might be a reason to test even though it is just passing the prop directly to the focus trap. It's your call.

My general guidance is to only test logic that is located in the exact file that you have imported to test (`PanelV2`). If the logic is located in another file in your project, then that file should be tested separately (`Portal`). If the logic comes from a third-party package or library, then it is their responsibility to test their own code (`focus-trap-react`) not yours.

jendowns replied:
Thanks that makes sense -- do you still think this test is worthwhile though, as-is? (And then something separate for focus trap, in the Portal tests most likely)

diego-codes replied:
I think it's good. I just realized that it tabs to the content area first, which makes sense why you'd want to tab through.

SimonFinney commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,103 @@
+        closeButton={{
+          label: 'test close',
+        }}
+        renderFooter={() => <Button>test footer button</Button>}
```
To be certain and until we remove them entirely, we should also validate that the `primaryButton` and `secondaryButton` props are not throwing a11y violations

SimonFinney commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,103 @@
+    expect(onCloseMock).toHaveBeenCalledTimes(1);
+  });
+
+  test('should not render panel when `isOpen` is set to `false`', () => {
```
👏

SimonFinney commented on src/components/PanelV2/__tests__/PanelV2.spec.js:
```diff
@@ -3,38 +3,103 @@
 describe('PanelV2', () => {
-  it('renders', () => {
-    const wrapper = mount(
+  test('should have no Axe or DAP violations', async () => {
```
I'm curious about the choice of `test` versus `it` here?

jendowns replied:
😆 I guess it go in my head while reading the React testing library docs & see all of their examples in that format. Though honestly I do kinda like `test` or `it` here, especially because I've flattened out the tests a bit. But do you have a preference?

(One more note actually -- it make help to do new structure with `test` to help us differentiate right away between new/re-evaluated tests and old tests?)

SimonFinney replied:
I've a feeling `it` is just an alias for familiarity between other testing libraries. I like that idea about differentiating!

jendowns commented:
@SimonFinney I believe I addressed your feedback above & this is ready to review again! Changes from #308 have been integrated as well, so there are no more DAP violations here.

SimonFinney commented:
:tada: This PR is included in version 1.16.0-prerelease.4 :tada:

The release is available on:
- [npm package (@canary dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security/v/1.16.0-prerelease.4)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v1.16.0-prerelease.4)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

SimonFinney commented:
:tada: This PR is included in version 1.16.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security/v/1.16.1)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v1.16.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

SimonFinney commented:
:tada: This PR is included in version 2.1.0 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security/v/2.1.0)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v2.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

SimonFinney commented:
:tada: This PR is included in version 1.16.1-prerelease.1 :tada:

The release is available on:
- [npm package (@canary dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security/v/1.16.1-prerelease.1)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v1.16.1-prerelease.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
